tololo/core/lib/carts/cart.ex

```diff
@@ -0,0 +1,135 @@
+defmodule TololoCore.Carts.Cart do
+  # @moduledoc """
+
+  # """
+  use Ash.Resource,
+    otp_app: :tololo,
+    domain: TololoCore.Carts,
+    extensions: [AshGraphql.Resource],
+    data_layer: AshPostgres.DataLayer,
+    authorizers: [Ash.Policy.Authorizer],
+    notifiers: [Ash.Notifier.PubSub, TololoCore.Kafka.AshNotifier]
+
+  graphql do
+    type :cart
+  end
+
+  postgres do
+    table "carts"
+    repo Tololo.Repo
+  end
+
+  field_policies do
+    field_policy :* do
+      description "the rest of the fields don't require any special policies"
+      authorize_if always()
+    end
+  end
+
+  code_interface do
+    define :create
+    define :add_variant, args: [:variant, {:optional, :quantity}, {:optional, :notes}]
+    define :checkout_delivery, args: [:cart, :delivery_input]
+  end
+
+  actions do
+    defaults [:read, :destroy, create: :*, update: :*]
+
+    action :checkout_delivery, :term do
+      argument :cart, :term, allow_nil?: false
+      argument :delivery_input, :map, allow_nil?: false
+
+      run fn %{arguments: %{cart: cart, delivery_input: delivery_input}}, _ ->
+        TololoCore.Deliveries.Delivery.initialize(
+          %{delivery_input | delivery_order: cart |> cart_to_map()},
+          authorize?: false
+        )
+      end
+    end
+
+    update :add_variant do
+      require_atomic? false
+      argument :variant, :struct, allow_nil?: false
+      argument :quantity, :integer, default: 1
+      argument :notes, :string, default: nil
+
+      validate fn %{data: %{currency: currency}, arguments: %{variant: %{prices: prices}}} =
+                    changeset,
+                  _context ->
+        prices = prices |> Ash.load!(:currency, lazy?: true, reuse_values?: true)
+
+        if prices |> Enum.any?(fn price -> price.currency == currency end) do
```

Comment:
If there is only one item in pipeline is best to not use a pipeline. And use the first param explicitly